packages/babel-plugin-transform-typescript/src/enum.ts

```diff
@@ -162,7 +162,26 @@ function ReferencedIdentifier(
 ) {
   const { seen, path, t } = state;
   const name = expr.node.name;
-  if (seen.has(name) && !expr.scope.hasOwnBinding(name)) {
+
+  if (seen.has(name)) {
+    for (
+      let curScope = expr.scope;
+      curScope !== path.scope;
+      curScope = curScope.parent
+    ) {
+      if (curScope.hasOwnBinding(name)) {
+        /* The name is declared inside enum:
+        enum Foo {
+          A,
+          B = (() => {
+            const A = 1;
+            return A;
+          })())
+        } */
+        return;
```

Comment:
This part seems like something we should provide in the scope prototype for Babel 8. The `Scope.prototype.hasBinding` could accept an `upToScope` parameter where the scope search will end.